### PR TITLE
[Bugfix:CourseMaterials] All course materials get deleted

### PR DESCRIPTION
**Pull request by akshatbatra**
### What is the current behavior?
Creating a folder with a particular name and then deleting it causes all course materials to get deleted.
Steps to reproduce:
1.) Add new course material by clicking on "Upload Course Materials" and typing in `%2e` (encoded form of `.`) or `%2e%2e` as the subdirectory name, select a file and click submit.
2.) Make sure that there are some files/folders other than this folder in order to verify the behavior.
2.) Delete the newly created folder named `%2e` or `%2e%2e`.
3.) See that all the course materials get removed.

### What is the new behavior?
A check has been added to prevent the deletion of the course materials folder.

Credit: Thanks @cjreed121 for advice on how to fix.

### CODE REVIEW

**Review by tkoz0 — COMMENTED** [review]
When a directory is created with the name `%2e`, it seems that there is no way to delete it. Deletion of the `course_materials` directory is prevented, but so is deletion of a directory named `%2e`.

**Review by tkoz0 — COMMENTED** [review]
Upon further investigation, other issues can happen due to decoding substrings such as `%2e` or `%2f`. For example,
1. Upload a file to a directory named `1%2f2` (decoded to `1/2`).
2. Upload a file to directory `1/2`.
3. Delete directory `1%2f2`
4. Observe that `1/2` is deleted but `1%2f2` remains.

**Comment by akshatbatra** [comment]
It has been identified that this issue is there for zip files such as [file.zip](https://github.com/Submitty/Submitty/files/9049579/file.zip), so, a check has been added to take care of that.
.

File: site/app/controllers/course/CourseMaterialsController.php [hunk 1]
```diff
@@ -114,8 +114,12 @@ public function deleteCourseMaterial($id) {
         }
         // security check
         $dir = "course_materials";
-        $path = $this->core->getAccess()->resolveDirPath($dir, htmlspecialchars_decode(rawurldecode($cm->getPath())));
-
+        $path = $this->core->getAccess()->resolveDirPath($dir, $cm->getPath());
+        // check to prevent the deletion of course_materials folder
+        if ($path === $this->core->getConfig()->getCoursePath() . "/uploads/course_materials") {
```
Comment (cjreed121) [inline site/app/controllers/course/CourseMaterialsController.php:119]: Rather than concatenating strings can you use FileUtils::joinPaths

**Review by cjreed121 — CHANGES_REQUESTED** [review]

**Review by cjreed121 — APPROVED** [review]

**Review by tkoz0 — APPROVED** [review]
Works as expected now.

**Pull request merged**